pydantic/aliases.py

```diff
@@ -0,0 +1,125 @@
+"""Support for alias configurations."""
+from __future__ import annotations
+
+import dataclasses
+from typing import Callable, Literal
+
+from ._internal import _internal_dataclass
+
+__all__ = ('AliasGenerator', 'AliasPath', 'AliasChoices')
+
+
+@dataclasses.dataclass(**_internal_dataclass.slots_true)
+class AliasPath:
+    """Usage docs: https://docs.pydantic.dev/2.6/concepts/alias#aliaspath-and-aliaschoices
+
+    A data class used by `validation_alias` as a convenience to create aliases.
+
+    Attributes:
+        path: A list of string or integer aliases.
+    """
+
+    path: list[int | str]
+
+    def __init__(self, first_arg: str, *args: str | int) -> None:
+        self.path = [first_arg] + list(args)
+
+    def convert_to_aliases(self) -> list[str | int]:
+        """Converts arguments to a list of string or integer aliases.
+
+        Returns:
+            The list of aliases.
+        """
+        return self.path
+
+
+@dataclasses.dataclass(**_internal_dataclass.slots_true)
+class AliasChoices:
+    """Usage docs: https://docs.pydantic.dev/2.6/concepts/alias#aliaspath-and-aliaschoices
+
+    A data class used by `validation_alias` as a convenience to create aliases.
+
+    Attributes:
+        choices: A list containing a string or `AliasPath`.
+    """
+
+    choices: list[str | AliasPath]
+
+    def __init__(self, first_choice: str | AliasPath, *choices: str | AliasPath) -> None:
+        self.choices = [first_choice] + list(choices)
+
+    def convert_to_aliases(self) -> list[list[str | int]]:
+        """Converts arguments to a list of lists containing string or integer aliases.
+
+        Returns:
+            The list of aliases.
+        """
+        aliases: list[list[str | int]] = []
+        for c in self.choices:
+            if isinstance(c, AliasPath):
+                aliases.append(c.convert_to_aliases())
+            else:
+                aliases.append([c])
+        return aliases
+
+
+@dataclasses.dataclass(**_internal_dataclass.slots_true)
+class AliasGenerator:
+    """Usage docs: https://docs.pydantic.dev/2.6/concepts/alias#using-an-aliasgenerator
+
+    A data class used by `alias_generator` as a convenience to create various aliases.
+
+    Attributes:
+        validation_alias: A callable that takes a field name and returns a validation alias for it.
+        alias: A callable that takes a field name and returns an alias for it.
+        serialization_alias: A callable that takes a field name and returns a serialization alias for it.
+    """
+
+    validation_alias: Callable[[str], str | AliasPath | AliasChoices] | None = None
+    alias: Callable[[str], str] | None = None
+    serialization_alias: Callable[[str], str] | None = None
```

Comment:
I assume we need all three to satisfy issues around prioritisation?

If so, please can we describe that somewhere (sorry if we have and I've missed it).

Otherwise, could we remove `alias` here?

Reply:
Speaking of the above use case, it doesn't seem to be documented anywhere that `AliasGenerator` can be used as `AliasGenerator(lambda x: x.upper())`. Might not be immediately clear to pydantic users unless  they view the __init__ signature.

Reply:
I'll go with the easy route of `alias` first 👍. @jusexton, I think given the removal of the more customized `__init__`, the behavior should be clear enough (as in, standard behavior for positional args and dataclasses)